### ISSUE TITLE
SCA doesn't send register modified alert when custom policy is defined

### ISSUE THREAD

**Issue by okynos** [wazuh/wazuh#2981]
Hi team,

We are doing some testing over `SCA` module, we have defined this policy file:
```
policy:
  id: "win_custom"
  file: "win_custom_rcl.yml"
  name: "Custom Benchmark for Windows registry"
  description: "This document provides a way to test SCA on Windows systems."

checks:
 - id: 1
   title: 'Register r1 enabled'
   compliance:
    - pci_dss: '10.6.1'
   condition: any
   rules:
     - 'r:HKCU\Software\test -> r1 -> 1;'
 - id: 2
   title: 'Register r2 enabled'
   compliance:
    - pci_dss: '10.6.1'
   condition: any
   rules:
     - 'r:HKCU\Software\test -> r2 -> 1;'
 - id: 3
   title: 'Register r3 enabled'
   compliance:
    - pci_dss: '10.6.1'
   condition: any
   rules:
     - 'r:HKCU\Software\test -> r3 -> 1;'
 - id: 4
   title: 'Register r4 enabled'
   compliance:
    - pci_dss: '10.6.1'
   condition: any
   rules:
     - 'r:HKCU\Software\test -> r4 -> 1;'
 - id: 5
   title: 'Register r5 enabled'
   compliance:
    - pci_dss: '10.6.1'
   condition: any
   rules:
     - 'r:HKCU\Software\test -> r5 -> 1;'
 - id: 6
   title: 'Register r6 enabled'
   compliance:
    - pci_dss: '10.6.1'
   condition: any
   rules:
     - 'r:HKCU\Software\test -> r6 -> 1;'
 - id: 7
   title: 'Register r7 enabled'
   compliance:
    - pci_dss: '10.6.1'
   condition: any
   rules:
     - 'r:HKCU\Software\test -> r7 -> 1;'
 - id: 8
   title: 'Register r8 enabled'
   compliance:
    - pci_dss: '10.6.1'
   condition: any
   rules:
     - 'r:HKCU\Software\test -> r8 -> 1;'
 - id: 9
   title: 'Register r9 enabled'
   compliance:
    - pci_dss: '10.6.1'
   condition: any
   rules:
     - 'r:HKCU\Software\test -> r9 -> 1;'
 - id: 10
   title: 'Register r10 enabled'
   compliance:
    - pci_dss: '10.6.1'
   condition: any
   rules:
     - 'r:HKCU\Software\test -> r10 -> 1;'
```

We have the following `ossec.conf`:
```
<ossec_config>
  <client>
    <server>
      <address>192.168.121.100</address>
      <port>1514</port>
      <protocol>udp</protocol>
    </server>
    <crypto_method>aes</crypto_method>
    <notify_time>10</notify_time>
    <time-reconnect>60</time-reconnect>
    <auto_restart>yes</auto_restart>
  </client>

  <sca>
    <enabled>yes</enabled>
    <scan_on_start>yes</scan_on_start>
    <interval>60s</interval>
    <skip_nfs>yes</skip_nfs>

    <policies>
      <policy>win_audit_rcl.yml</policy>
      <policy>win_custom_rcl.yml</policy>
    </policies>
  </sca>

  <logging>
    <log_format>plain</log_format>
  </logging>
</ossec_config>
```

And we have a script to create and modify registers on windows like this:
```
$originalLimit=10
$customRegistryPath="HKCU:\Software\test"

New-Item $customRegistryPath -Force
while($true)
{
  For ($i=0; $i -le $originalLimit; $i++) {
    New-ItemProperty -Path $customRegistryPath -Name "r$i" -Value 1 -Force
  }
  Start-Sleep -Seconds 7

  For ($i=0; $i -le $originalLimit; $i++) {
    New-ItemProperty -Path $customRegistryPath -Name "r$i" -Value 0 -Force
  }
  Start-Sleep -Seconds 7
}
```

We received with success the first alert of first scan of SCA but the successive alerts with changed registers aren't received on manager side, a part of debug log is:
```
2019/03/29 15:51:46 sca[876] wm_sca.c:487 at wm_sca_read_files(): INFO: Starting evaluation of policy: 'win_custom_rcl.yml
2019/03/29 15:51:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r1 enabled'.
2019/03/29 15:51:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r1 -> 1;
2019/03/29 15:51:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:51:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:51:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r2 enabled'.
2019/03/29 15:51:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r2 -> 1;
2019/03/29 15:51:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:51:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:51:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r3 enabled'.
2019/03/29 15:51:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r3 -> 1;
2019/03/29 15:51:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:51:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:51:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r4 enabled'.
2019/03/29 15:51:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r4 -> 1;
2019/03/29 15:51:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:51:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:51:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r5 enabled'.
2019/03/29 15:51:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r5 -> 1;
2019/03/29 15:51:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:51:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:51:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r6 enabled'.
2019/03/29 15:51:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r6 -> 1;
2019/03/29 15:51:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:51:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:51:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r7 enabled'.
2019/03/29 15:51:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r7 -> 1;
2019/03/29 15:51:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:51:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:51:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r8 enabled'.
2019/03/29 15:51:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r8 -> 1;
2019/03/29 15:51:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:51:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:51:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r9 enabled'.
2019/03/29 15:51:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r9 -> 1;
2019/03/29 15:51:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:51:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:51:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r10 enabled'.
2019/03/29 15:51:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r10 -> 1;
2019/03/29 15:51:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:51:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:51:46 sca[876] wm_sca.c:492 at wm_sca_read_files(): DEBUG: Calculating hash for scanned results.
2019/03/29 15:51:49 sca[876] wm_sca.c:1943 at wm_sca_send_summary(): DEBUG: Sending summary event for file: 'win_custom_rcl.yml
2019/03/29 15:51:49 sca[876] wm_sca.c:192 at wm_sca_send_alert(): DEBUG: Sending event: {"type":"summary","scan_id":594930025,"name":"Custom Benchmark for Windows registry","policy_id":"win_custom","file":"win_custom_rcl.yml","description":"This document provides a wa
y to test SCA on Windows systems.","passed":10,"failed":0,"score":100,"start_time":1553874706,"end_time":1553874706,"hash":"73b6ce4469d59286213719fa06508458"}
2019/03/29 15:51:49 sca[876] wm_sca.c:504 at wm_sca_read_files(): INFO: Evaluation finished for policy 'win_custom_rcl.yml'.
2019/03/29 15:51:49 sca[876] wm_sca.c:232 at wm_sca_send_policies_scanned(): DEBUG: Sending scanned policies.
2019/03/29 15:51:49 sca[876] wm_sca.c:192 at wm_sca_send_alert(): DEBUG: Sending event: {"type":"policies","policies":["win_audit","win_custom"]}
2019/03/29 15:51:50 sca[876] wm_sca.c:304 at wm_sca_start(): INFO: Security Configuration Assessment scan finished. Duration: 7 seconds.
2019/03/29 15:51:50 sca[876] wm_sca.c:344 at wm_sca_start(): DEBUG: Sleeping for 53 seconds
2019/03/29 15:52:43 sca[876] wm_sca.c:293 at wm_sca_start(): INFO: Starting Security Configuration Assessment scan.
2019/03/29 15:52:43 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Check for Windows platform'.
2019/03/29 15:52:43 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKEY_LOCAL_MACHINE\SAM\SAM;
2019/03/29 15:52:43 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKEY_LOCAL_MACHINE\SAM\SAM'.
2019/03/29 15:52:43 sca[876] wm_sca.c:868 at wm_sca_do_scan(): DEBUG: Found registry.
2019/03/29 15:52:43 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.

...

2019/03/29 15:52:46 sca[876] wm_sca.c:487 at wm_sca_read_files(): INFO: Starting evaluation of policy: 'win_custom_rcl.yml
2019/03/29 15:52:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r1 enabled'.
2019/03/29 15:52:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r1 -> 1;
2019/03/29 15:52:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:52:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:52:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r2 enabled'.
2019/03/29 15:52:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r2 -> 1;
2019/03/29 15:52:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:52:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:52:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r3 enabled'.
2019/03/29 15:52:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r3 -> 1;
2019/03/29 15:52:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:52:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:52:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r4 enabled'.
2019/03/29 15:52:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r4 -> 1;
2019/03/29 15:52:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:52:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:52:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r5 enabled'.
2019/03/29 15:52:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r5 -> 1;
2019/03/29 15:52:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:52:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:52:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r6 enabled'.
2019/03/29 15:52:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r6 -> 1;
2019/03/29 15:52:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:52:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:52:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r7 enabled'.
2019/03/29 15:52:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r7 -> 1;
2019/03/29 15:52:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:52:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:52:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r8 enabled'.
2019/03/29 15:52:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r8 -> 1;
2019/03/29 15:52:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:52:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:52:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r9 enabled'.
2019/03/29 15:52:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r9 -> 1;
2019/03/29 15:52:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:52:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:52:46 sca[876] wm_sca.c:766 at wm_sca_do_scan(): DEBUG: Checking entry: 'Register r10 enabled'.
2019/03/29 15:52:46 sca[876] wm_sca.c:778 at wm_sca_do_scan(): DEBUG: Rule is: r:HKCU\Software\test -> r10 -> 1;
2019/03/29 15:52:46 sca[876] wm_sca.c:866 at wm_sca_do_scan(): DEBUG: Checking registry: 'HKCU\Software\test'.
2019/03/29 15:52:46 sca[876] wm_sca.c:1006 at wm_sca_do_scan(): DEBUG: Condition ANY.
2019/03/29 15:52:46 sca[876] wm_sca.c:492 at wm_sca_read_files(): DEBUG: Calculating hash for scanned results.
2019/03/29 15:52:49 sca[876] wm_sca.c:1943 at wm_sca_send_summary(): DEBUG: Sending summary event for file: 'win_custom_rcl.yml
2019/03/29 15:52:49 sca[876] wm_sca.c:192 at wm_sca_send_alert(): DEBUG: Sending event: {"type":"summary","scan_id":1194131157,"name":"Custom Benchmark for Windows registry","policy_id":"win_custom","file":"win_custom_rcl.yml","description":"This document provides a w
ay to test SCA on Windows systems.","passed":10,"failed":0,"score":100,"start_time":1553874766,"end_time":1553874766,"hash":"73b6ce4469d59286213719fa06508458"}
2019/03/29 15:52:49 sca[876] wm_sca.c:504 at wm_sca_read_files(): INFO: Evaluation finished for policy 'win_custom_rcl.yml'.
2019/03/29 15:52:49 sca[876] wm_sca.c:232 at wm_sca_send_policies_scanned(): DEBUG: Sending scanned policies.
2019/03/29 15:52:49 sca[876] wm_sca.c:192 at wm_sca_send_alert(): DEBUG: Sending event: {"type":"policies","policies":["win_audit","win_custom"]}
2019/03/29 15:52:50 sca[876] wm_sca.c:304 at wm_sca_start(): INFO: Security Configuration Assessment scan finished. Duration: 7 seconds.
2019/03/29 15:52:50 sca[876] wm_sca.c:344 at wm_sca_start(): DEBUG: Sleeping for 53 seconds
```

|Wazuh version|Install type|Install method|Platform|
|---|---|---|---|
| 3.9.0-3910 | Agent | Packages | Windows Server 2016 |

Regards.

**Comment by chemamartinez** [comment]
It seems it is a privileges issue with some registries that the module is unable to read. Two steps are needed here:

- Check that those registries unable to read for SCA, can be monitor by Syscheck. If Syscheck can read that keys, it is clear it is an SCA related bug.

- If it is a privileges issue, we should improve the module logging to know what is happening with those registries.

**Comment by JuantAldea** [comment]
@okynos As which user are you editing the windows registry?
ossec-agent runs as SYSTEM user, if you are altering the registry as any other user other than SYSTEM, ossec-agent won't see this change, has HKCU is user-specific

**Comment by okynos** [comment]
Hi @JuantAldea,

It's true that executing script with `psexec` do the trick but now on `regedit` we can't see our registers and instead we can't manually modify in an supposed case also, it seems that `psexec` is specific development tool that doesn't come natively on any Windows system, there's no way to do on `Administrator` account or user account like normal user does?, thanks.

Regards.

**Comment by chemamartinez** [comment]
Merged: #3052.
